### PR TITLE
Add agenthealth extension with user agent handler.

### PR DESCRIPTION
**Pull request by jefchien**
# Description of the issue
The agenthealth extension implements the `awsmiddleware.Extension` interface and provides the request/response handlers to configure on the AWS Clients. This allows the agent's user agent to be prepended on AWS requests.

# Description of changes
Adds the agenthealth extension scaffolding with a user agent handler. There's a singleton user agent that gets updated by its `Set*` functions. On each change, the handler is updated and on each requests, the user agent string is added.

# License
By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

# Tests
Adds unit tests.

# Requirements
_Before commit the code, please do the following steps._
1. Run `make fmt` and `make fmt-sh`
2. Run `make lint`

### CODE REVIEW

File: extension/agenthealth/extension.go [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+	component.StartFunc
+	component.ShutdownFunc
+}
+
+var _ awsmiddleware.Extension = (*agentHealth)(nil)
+
+func (ah *agentHealth) Handlers() ([]awsmiddleware.RequestHandler, []awsmiddleware.ResponseHandler) {
+	var responsesHandlers []awsmiddleware.ResponseHandler
+	requestsHandlers := []awsmiddleware.RequestHandler{useragent.NewHandler(ah.cfg.IsUsageDataEnabled)}
+	return requestsHandlers, responsesHandlers
```
Comment (lisguo) [inline extension/agenthealth/extension.go:26]: nit: sounds weird plural. Should it just be `requestHandlers` and `responseHandlers`?

File: extension/agenthealth/handler/useragent/handler.go [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+	return handlerID
+}
+
+func (uah *userAgentHandler) Position() awsmiddleware.HandlerPosition {
+	return awsmiddleware.After
+}
+
+// HandleRequest prepends the User-Agent header with the CloudWatch Agent's
+// user agent string.
+func (uah *userAgentHandler) HandleRequest(_ string, r *http.Request) {
```
Comment (lisguo) [inline extension/agenthealth/handler/useragent/handler.go:31]: is there a future implementation of the string param?

File: extension/agenthealth/handler/useragent/useragent.go [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+	inputsStr     atomic.String
+	processorsStr atomic.String
+	outputsStr    atomic.String
+}
+
+var _ UserAgent = (*userAgent)(nil)
+
+func (ua *userAgent) SetComponents(otelCfg *otelcol.Config, telegrafCfg *telegraf.Config) {
+	ua.dataLock.Lock()
+	defer ua.dataLock.Unlock()
```
Comment (lisguo) [inline extension/agenthealth/handler/useragent/useragent.go:74]: how likely are we going to run into concurrent reads here? We only call `SetComponents` once on init right?

File: extension/agenthealth/handler/useragent/useragent.go [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+				}
+				if cfg.IsEnhancedContainerInsights() {
+					ua.outputs.Add(flagEnhancedContainerInsights)
+				}
+			}
+		}
+	}
+
+	if !ua.isRoot {
+		ua.inputs.Add(flagRunAsUser)
```
Comment (lisguo) [inline extension/agenthealth/handler/useragent/useragent.go:106]: interesting, are we trying to see how many customers run as non root?

**Review by lisguo — DISMISSED** [review]

File: extension/agenthealth/extension.go (hunk 1 above)
Comment (jefchien) [reply extension/agenthealth/extension.go:26]: Yep it's a typo.

File: extension/agenthealth/handler/useragent/handler.go (hunk 2 above)
Comment (jefchien) [reply extension/agenthealth/handler/useragent/handler.go:31]: The user agent handler doesn't need it, but other ones, like the client stats one (measuring latency) do.

File: extension/agenthealth/handler/useragent/useragent.go (hunk 3 above)
Comment (jefchien) [reply extension/agenthealth/handler/useragent/useragent.go:74]: Yeah you're right. I remember thinking that and then forgetting to remove it.

File: extension/agenthealth/handler/useragent/useragent.go (hunk 4 above)
Comment (jefchien) [reply extension/agenthealth/handler/useragent/useragent.go:106]: This is an existing flag we set in the user agent.

File: go.mod [hunk 5]
```diff
@@ -6,7 +6,7 @@ replace github.com/influxdata/telegraf => github.com/aws/telegraf v0.10.2-0.2022
 
 // Replace with https://github.com/amazon-contributing/opentelemetry-collector-contrib, there are no requirements for all receivers/processors/exporters
 // to be all replaced since there are some changes that will always be from upstream
-replace github.com/open-telemetry/opentelemetry-collector-contrib/exporter/awsemfexporter => github.com/amazon-contributing/opentelemetry-collector-contrib/exporter/awsemfexporter v0.0.0-20231020163023-8bdf732d320a
```
Comment (jefchien) [inline go.mod:9]: Will rebase this when we pull in the contrib changes.

**Review by bryce-carey — DISMISSED** [review]

**Review by bryce-carey — APPROVED** [review]

**Review by SaxyPandaBear — APPROVED** [review]

**Pull request merged**